s3/s3.go

```diff
@@ -0,0 +1,243 @@
+	if !ok {
+		fmt.Println("Invalid prefix format.")
+		return false
+	}
+
+	// Set level, filename and key
+	var level, filename, key string
+	if argv["put"] == true {
+		level = "read-write"
+		filename, ok = argv["<file>"].(string)
+		if !ok {
+			fmt.Println("Invalid file format.")
+			return false
+		}
+	} else if argv["get"] == true {
+		level = "read-write" // can change this later
+		key, ok = argv["<key>"].(string)
+		if !ok {
+			fmt.Println("Invalid key format.")
+			return false
+		}
+	}
+
+	// Load configuration
+	config, err := config.Load()
+	if err != nil {
+		fmt.Println("Failed to load configuration file, error: ", err)
+		return false
+	}
+
+	// Set credentials for auth
+	myAuth := auth.New(
+		&tcclient.Credentials{
+			ClientID:    config["config"]["clientId"].(string),
+			AccessToken: config["config"]["accessToken"].(string),
+			Certificate: config["config"]["certificate"].(string),
+		},
+	)
+
+	// Get credentials for AWS S3
+	resp, err := myAuth.AwsS3Credentials(level, bucket, prefix)
+	if err != nil {
+		fmt.Println("Failed to load AWS S3 credentials: ", err)
+		return false
+	}
+
+	// Set Credentials for AWS S3
+	aws_access_key_id := resp.Credentials.AccessKeyID
+	aws_secret_access_key := resp.Credentials.SecretAccessKey
+	aws_session_token := resp.Credentials.SessionToken
+
+	creds := credentials.NewStaticCredentials(aws_access_key_id, aws_secret_access_key, aws_session_token)
+	_, err = creds.Get()
+	if err != nil {
+		fmt.Printf("Invalid credentials: %s", err)
+		return false
+	}
+
+	// Get bucket location
+	aws_config := aws.NewConfig().WithCredentials(creds).WithRegion("us-west-2")
+	sess := session.New(aws_config)
+	svc := awsS3.New(sess)
+	params := &awsS3.GetBucketLocationInput{
+		Bucket: aws.String(bucket),
+	}
+	res, err := svc.GetBucketLocation(params)
+	if err != nil {
+		fmt.Println(err.Error())
+		return false
+	}
+	region := res.LocationConstraint
+
+	// Create a session which contains the configurations for the SDK.
+	// Use the session to create the service clients to make API calls to AWS.
+	aws_config = aws.NewConfig().WithCredentials(creds).WithRegion(*region)
+	sess = session.New(aws_config)
+	svc = awsS3.New(sess)
+
+	if argv["put"] == true {
+		upload := Put(svc, filename, bucket, prefix)
+		return upload
+	} else if argv["get"] == true {
+		download := Get(svc, key, bucket, prefix)
+		return download
+	} else {
+		panic(fmt.Sprint("Invalid subcommand."))
+	}
+}
+
+func Put(svc *awsS3.S3, filename string, bucket string, prefix string) bool {
+	// Open the filename and return the file
+	file, err := os.Open(filename)
+	if err != nil {
+		fmt.Println("Failed to open file", filename, err)
+		return false
+	}
+	defer file.Close()
+
+	// Generate key from filename
+	key := filepath.Base(filename)
+
+	fmt.Printf("Uploading %s to s3://%s/%s...\n", key, bucket, prefix)
+
+	// Upload the file
+	params := &awsS3.PutObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(prefix + key),
```

Comment:
hmm... I didn't think about the details here...

Maybe we should do it, such that we get:
```
$ taskcluster upload myfile.txt mybucket some/where             # myfile.txt ->  some/where
$ taskcluster upload myfile.txt mybucket some/where/myfile.txt  # myfile.txt -> some/where/myfile.txt
$ taskcluster upload myfile.txt mybucket some/where/            # myfile.txt -> some/where/myfile.txt
$ taskcluster upload myfile.txt mybucket some/where.txt         # myfile.txt -> some/where.txt
$ taskcluster upload myfile.txt mybucket some/where.txt/        # myfile.txt -> some/where.txt/myfile.txt
```

Ie. if the `prefix` ends in a slash then we add the `filepath.Base(filename)` otherwise, we don't add anything...
Similar to how `cp` works on linux... Sounds reasonable?